aptos-move/framework/aptos-stdlib/sources/pool_u64.move

```diff
@@ -0,0 +1,188 @@
+    use aptos_std::simple_map::{Self, SimpleMap};
+    use std::error;
+    use std::vector;
+
+    /// Shareholder not present in pool.
+    const ESHAREHOLDER_NOT_FOUND: u64 = 1;
+    /// There are too many shareholders in the pool.
+    const ETOO_MANY_SHAREHOLDERS: u64 = 2;
+    /// Cannot destroy non-empty pool.
+    const EPOOL_IS_NOT_EMPTY: u64 = 3;
+    /// Cannot redeem more shares than the shareholder has in the pool.
+    const EINSUFFICIENT_SHARES: u64 = 4;
+
+    struct Pool has store {
+        shareholders_limit: u64,
+        // We store numbers internally as u128 to make sure u64 multiplications don't overflow.
+        total_coins: u128,
+        total_shares: u128,
+        shares: SimpleMap<address, u128>,
+        shareholders: vector<address>,
+    }
+
+    /// Create a new pool.
+    public fun create(shareholders_limit: u64): Pool {
+        Pool {
+            shareholders_limit,
+            total_coins: 0,
+            total_shares: 0,
+            shares: simple_map::create<address, u128>(),
+            shareholders: vector::empty<address>(),
+        }
+    }
+
+    /// Destroy an empty pool. This will fail if the pool has any balance of coins.
+    public fun destroy_empty(pool: Pool) {
+        assert!(pool.total_coins == 0, error::invalid_state(EPOOL_IS_NOT_EMPTY));
+        let Pool {
+            shareholders_limit: _,
+            total_coins: _,
+            total_shares: _,
+            shares: _,
+            shareholders: _,
+        } = pool;
+    }
+
+    /// Return `pool`'s total balance of coins.
+    public fun total_coins(pool: &Pool): u64 {
+        (pool.total_coins as u64)
+    }
+
+    /// Return the total number of shares across all shareholders in `pool`.
+    public fun total_shares(pool: &Pool): u64 {
+        (pool.total_shares as u64)
+    }
+
+    /// Return true if `shareholder` is in `pool`.
+    public fun contains(pool: &Pool, shareholder: address): bool {
+        simple_map::contains_key(&pool.shares, &shareholder)
+    }
+
+    /// Return the number of shares of `stakeholder` in `pool`.
+    public fun shares(pool: &Pool, shareholder: address): u64 {
+        if (contains(pool, shareholder)) {
+            (*simple_map::borrow(&pool.shares, &shareholder) as u64)
+        } else {
+            0
+        }
+    }
+
+    /// Return the balance in coins of `shareholder` in `pool.`
+    public fun balance(pool: &Pool, shareholder: address): u64 {
+        let num_shares = shares(pool, shareholder);
+        shares_to_amount(pool, num_shares)
+    }
+
+    /// Return the list of shareholders in `pool`.
+    public fun shareholders(pool: &Pool): vector<address> {
+        pool.shareholders
+    }
+
+    /// Return the number of shareholders in `pool`.
+    public fun shareholders_count(pool: &Pool): u64 {
+        vector::length(&pool.shareholders)
+    }
+
+    /// Update `pool`'s total balance of coins.
+    public fun update_total_coins(pool: &mut Pool, new_total_coins: u64) {
```

Comment:
maybe add coins not update, this should be purely incremental? or can we at least evaluate that this is increasing?

Reply:
There can be use cases where the total coins decrease. For example, if there's a penalty such as stake slashing.